Assets/MixedRealityToolkit.Extensions/SpectatorView/Scripts/Calibration/HeadsetCalibration.cs

```diff
@@ -0,0 +1,267 @@
+﻿using System.Collections.Generic;
+using UnityEngine;
+
+using Microsoft.MixedReality.Toolkit.Extensions.Experimental.MarkerDetection;
+using Microsoft.MixedReality.Toolkit.Extensions.Experimental.SpectatorView.MarkerDetection;
+using Microsoft.MixedReality.Toolkit.Extensions.Experimental.SpectatorView.Utilities;
+using Microsoft.MixedReality.Toolkit.Extensions.PhotoCapture;
+using System.Text;
+using System;
+using System.Collections.Concurrent;
+
+namespace Microsoft.MixedReality.Toolkit.Extensions.Experimental.SpectatorView
+{
+    public delegate void HeadsetCalibrationDataUpdatedHandler(byte[] data);
+
+    public class HeadsetCalibration : MonoBehaviour
+    {
+        /// <summary>
+        /// Check to show debug visuals for the detected markers.
+        /// </summary>
+        [Tooltip("Check to show debug visuals for the detected markers.")]
+        [SerializeField]
+        protected bool showDebugVisuals = true;
+
+        /// <summary>
+        /// QR Code Marker Detector in scene
+        /// </summary>
+        [Tooltip("QR Code Marker Detector in scene")]
+        [SerializeField]
+        protected QRCodeMarkerDetector qrCodeMarkerDetector;
+
+        /// <summary>
+        /// Debug Visual Helper in scene that will place game objects on qr code markers in the scene.
+        /// </summary>
+        [Tooltip("Debug Visual Helper in scene that will place game objects on qr code markers in the scene.")]
+        [SerializeField]
+        protected DebugVisualHelper qrCodeDebugVisualHelper;
+
+        /// <summary>
+        /// Debug Visual Helper in scene that will place game objects on aruco markers in the scene.
+        /// </summary>
+        [Tooltip("Debug Visual Helper in scene that will place game objects on aruco markers in the scene.")]
+        [SerializeField]
+        protected DebugVisualHelper arucoDebugVisualHelper;
+
+        private bool markersUpdated = false;
+        private Dictionary<int, Marker> qrCodeMarkers = new Dictionary<int, Marker>();
+        private Dictionary<int, GameObject> qrCodeDebugVisuals = new Dictionary<int, GameObject>();
+        private Dictionary<int, GameObject> arucoDebugVisuals = new Dictionary<int, GameObject>();
+        private readonly float markerPaddingRatio = 34f / (300f - (2f * 34f)); // padding pixels / marker width in pixels
+        private Dictionary<int, MarkerCorners> qrCodeMarkerCorners = new Dictionary<int, MarkerCorners>();
+        private Dictionary<int, MarkerCorners> arucoMarkerCorners = new Dictionary<int, MarkerCorners>();
```

Comment:
This is a nit and now that it's been written feel free to wontfix/ignore this - in general if we're gonna be storing multiple things on the same key, we should probably just store all of the things associated with that key on a sturct thing - so that we don't have to do 2-3 different hashes to do the samething (see lines 66-67 for where we are doing multiple hashes).